src/java_bytecode/java_bytecode_convert_method.cpp

```diff
@@ -972,109 +972,277 @@ bool java_bytecode_convert_methodt::class_needs_clinit(
+///
+/// This wrapper is called immediately before any read or write of a static
+/// field or whenever an instance of the class is created.
+/// When class \p classname does not need static initialization we
+/// return a nil irep.  We also explicitely avoid introducing calls to the
+/// wrapper for class C from the class initializer of class C.
+/// When class \p classname needs static initalization, this function creates a
+/// new function in the symbol table (or returns it if it had already been
+/// created on previous call) that implements (a simplification of) the
+/// algorithm defined in Section 5.5 of the JVM Specs. This function, or
+/// wrapper, checks whether static init has already taken place, calls the
+/// actual `<clinit>` method if not, and initializes super-classes and
+/// interfaces.
+///
+/// Note: The current implementation does not deal with exceptions.
+/// Note': was_ready is an optimization to make it clear to the symex engine
+/// that it doesn't need to unwind infinitely many copies of the wrapper.
+///
+/// The generated function looks like this:
+/// \code
+/// ```
+///   bool was_ready;
+///   if(C::__CPROVER_clinit_local_state == RETURN)
+///     return;
+///   C::__CPROVER_clinit_local_state=RETURN; // execute this only once!
+///
+///   ATOMIC_BEGIN
+///   Assume(C::__CPROVER_clinit_state != IN_PROGRESS)
+///   If(C::__CPROVER_clinit_state==NOT_INIT) {
+///     C::__CPROVER_clinit_state=IN_PROGRESS
+///     was_ready=false;
+///   }
+///   else If(C::__CPROVER_clinit_state==READY) {
+///     was_ready=true;
+///   }
+///   ATOMIC_END
+///
+///   if(was_ready)
+///     return;
+///
+///   C'::clinit();  // C' = superclass of C
+///   l_1::clinit()
+///   ...
+///   l_n::clinit()
+///
+///   C::clinit();
+///   ATOMIC_START
+///   C::__CPROVER_clinit_state=RETURN;
+///   ATOMIC_END
+///   return;
+/// ```
+/// \endcode
+///
+/// \param classname: The name of the class to initialize.
+/// \param caller_method_name: The fully qualified name (including descriptor)
+///   of the method in whose body this call will be added.
 /// \return Returns a symbol_exprt pointing to the given class' clinit wrapper
 ///   if one is required, or nil otherwise.
 exprt java_bytecode_convert_methodt::get_or_create_clinit_wrapper(
-  const irep_idt &classname)
+  const irep_idt &classname,
+  const irep_idt &caller_method_name)
 {
+  // decide if this class needs a class initializer wrapper at all, which
+  // requires finding out if it has a `<clinit>` method and whether any of the
+  // super classes do have it
   if(!class_needs_clinit(classname))
     return static_cast<const exprt &>(get_nil_irep());
 
-  // if the symbol table already contains the clinit_wrapper() function, return
-  // it
+  // if we are initializing class C due to a use of class C within method
+  // `C.<clinit>` then avoid the recursive (and unnecessary) call to
+  // the wrapper.
+  const irep_idt &real_clinit_name=id2string(classname)+".<clinit>:()V";
```

Comment:
Since this is new code, use the new clang-format-compatible / human-readable operator spacing (i.e. spaces around ==, =, etc)